src/thunderbots/software/multithreading/thread_safe_buffer.h

```diff
@@ -0,0 +1,157 @@
+     * - the destructor of this class is called
+     *
+     * @param max_wait_time The maximum duration to wait for a new value before
+     *                      returning
+     *
+     * @return The most recently value added to the buffer, or std::nullopt if none is
+     *         available
+     */
+    std::optional<T> pullMostRecentlyAddedValue(
+        Duration max_wait_time = Duration::fromSeconds(0));
+
+    /**
+     * Push the given value onto the buffer
+     *
+     * If the buffer is already full, this will overwrite the least recently added value
+     *
+     * @param value The value to push onto the buffer
+     */
+    void push(const T& value);
+
+    ~ThreadSafeBuffer();
+
+   private:
+    /**
+     * Waits for the buffer to have at least one value
+     *
+     * @param max_wait_time The maximum duration to wait for a new value before
+     *                      returning
+     *
+     * @return A unique_lock on the contents of the buffer
+     */
+    std::unique_lock<std::mutex> waitForBufferToHaveAValue(Duration max_wait_time);
+
+    std::mutex buffer_mutex;
+    boost::circular_buffer<T> buffer;
+
+    std::condition_variable received_new_value;
+
+    bool destructor_called;
+};
+
+template <typename T>
+ThreadSafeBuffer<T>::ThreadSafeBuffer(std::size_t buffer_size)
+    : buffer(buffer_size), destructor_called(false)
+{
+}
+
+template <typename T>
+std::optional<T> ThreadSafeBuffer<T>::pullLeastRecentlyAddedValue(Duration max_wait_time)
+{
+    auto buffer_lock = waitForBufferToHaveAValue(max_wait_time);
+
+    std::optional<T> result = std::nullopt;
+    if (!buffer.empty())
+    {
+        result = buffer.front();
+        buffer.pop_front();
+    }
+    return result;
+}
+
+template <typename T>
+std::optional<T> ThreadSafeBuffer<T>::pullMostRecentlyAddedValue(Duration max_wait_time)
+{
+    auto buffer_lock = waitForBufferToHaveAValue(max_wait_time);
+
+    std::optional<T> result = std::nullopt;
+    if (!buffer.empty())
+    {
+        result = buffer.back();
+        buffer.pop_back();
+    }
+    return result;
+}
+
+template <typename T>
+void ThreadSafeBuffer<T>::push(const T& value)
+{
+    std::scoped_lock<std::mutex> buffer_lock(buffer_mutex);
+    buffer.push_back(value);
+    received_new_value.notify_all();
+}
+
+template <typename T>
+std::unique_lock<std::mutex> ThreadSafeBuffer<T>::waitForBufferToHaveAValue(
+    Duration max_wait_time)
+{
+    std::unique_lock<std::mutex> buffer_lock(buffer_mutex);
+    received_new_value.wait_for(buffer_lock,
+                                std::chrono::duration<float>(max_wait_time.getSeconds()),
+                                [this] { return !buffer.empty() || destructor_called; });
+
+    // NOTE: We need to return this in order to prevent it being destructed so
+    //       the the lock is maintained until the value is read
+    return buffer_lock;
+}
+
+template <typename T>
+ThreadSafeBuffer<T>::~ThreadSafeBuffer()
+{
+    destructor_called = true;
```

Comment:
An interesting thought is whether or not this actually is a race condition, solely because it's a single bit and because of how we access it.

Regardless, I agree, we should be explicit, and better safe then sorry. Added (and used) a `in_destructor_mutex`

Reply:
Yea let's not risk it and play games with threads :stuck_out_tongue: